src/python/pants/backend/nfpm/fields/contents.py

```diff
@@ -0,0 +1,655 @@
+        if value is not None and value > 0o7777:
+            raise InvalidFieldException(
+                f"The '{cls.alias} field in target {address} must be less than or equal to "
+                f"0o7777, but was set to {repr(raw_value)} (ie: `{value:#o}`)."
+            )
+        return value
+
+
+class NfpmContentFileMtimeField(StringField):
+    nfpm_alias = "contents.[].file_info.mtime"
+    alias: ClassVar[str] = "file_mtime"
+    default = mtime_default
+    help = help_text(
+        f"""
+        The file modification time as an RFC 3339 formatted string.
+
+        For example: 2008-01-02T15:04:05Z
+
+        The format is defined in RFC 3339: https://rfc-editor.org/rfc/rfc3339.html
+
+        Though nFPM supports pulling mtime from the src file or directory in most
+        cases, the pants nfpm backend does not support this. Reading the mtime from
+        the filesystem is problematic because Pants does not track the mtime of files
+        and does not propagate any file mtime into the sandboxes. Reasons for this
+        include: git does not track mtime, timestamps like mtime cause many issues
+        for reproducible packaging builds, and reproducible builds are required
+        for pants to provide its fine-grained caches.
+
+        The default value is {repr(mtime_default)}. You may also override
+        the default value by setting `[nfpm].default_mtime` in `pants.toml`,
+        or by setting the `SOURCE_DATE_EPOCH` environment variable.
+
+        See also: https://reproducible-builds.org/docs/timestamps/
+        """
+    )
+
+    def normalized_value(self, default_mtime: str | None):
+        if self.value == self.default and default_mtime and self.value != default_mtime:
+            return default_mtime
+        return self.value
+
+
+# -----------------------------------------------------------------------------------------------
+# Internal generic parent class fields
+# -----------------------------------------------------------------------------------------------
+
+
+class InvalidFieldMemberTypeException(InvalidFieldException):
+    # based on InvalidFieldTypeException
+    def __init__(
+        self,
+        address: Address,
+        field_alias: str,
+        raw_value: Optional[Any],
+        *,
+        expected_type: str,
+        at_index: int,
+        wrong_element: Any,
+        description_of_origin: str | None = None,
+    ) -> None:
+        super().__init__(
+            softwrap(
+                f"""
+                The {repr(field_alias)} field in target {address} must be an iterable with
+                elements that have type {expected_type}. Encountered the element `{wrong_element}`
+                of type {type(wrong_element)} instead of {expected_type} at index {at_index}:
+                `{repr(raw_value)}`
+                """
+            ),
+            description_of_origin=description_of_origin,
+        )
+
+
+class _SrcDstSequenceField(Field):
```

Comment:
Should this be moved in a core Pants package?

Reply:
That works for me. Move it as `StringStringTupleSequenceField`, or can you think of a better name? And I think it makes sense to leave the dst dupes detection here.

Reply:
I made a more generic `TupleSequenceField`.